File: terrabutler/tf.py
import os
import signal
import subprocess
from colorama import Fore
from sys import exit
from terrabutler.settings import get_settings
from terrabutler.utils import paths

# Values from Config
org = get_settings()["general"]["organization"]


def terraform_args_print(command, site):
    """
    Print Args
    """

    if command == "init":
        needed_options = "backend"
    elif command == "plan" or command == "apply":
        needed_options = "var"
    else:
        needed_options = ""

    options = terraform_needed_options_builder(needed_options, site)
    return " ".join(options)


def terraform_needed_options_builder(needed_options, site):
    """
    Create array of needed options for backend or var files
    """
    from terrabutler.env import get_current_env
    env = get_current_env()
    default_env = get_settings()["environments"]["default"]["name"]

    if needed_options == "backend":
        backend_dir = paths["backends"]

        if site == "inception":  # Init inception with default ENV
            return ["-backend-config",
                    f"{backend_dir}/{org}-{default_env}-inception.tfvars"]
        else:
            return ["-backend-config",
                    f"{backend_dir}/{org}-{env}-{site}.tfvars"]

    elif needed_options == "var":
        variables_dir = paths["variables"]

        return ["-var-file", f"{variables_dir}/global.tfvars",
                "-var-file", f"{variables_dir}/{org}-{env}.tfvars",
                "-var-file", f"{variables_dir}/{org}-{env}-{site}.tfvars"
                ]

    else:  # If needed_options is empty, return empty array
        return []


def terraform_command_builder(command, site, args=[], options=[],
                              needed_options=""):
    """
    Create the command to run terraform
    """
    aux = ["terraform", command]  # Start base command

    if needed_options == "backend" or needed_options == "var":
        aux += terraform_needed_options_builder(needed_options, site)

    aux += options  # Add options passed by user
    aux += args  # Add args passed by user

    return aux


def terraform_command_runner(command, site, args=[], options=[],
                             needed_options=""):
    """
    Run tfenv and run the terraform command
    """
    from terrabutler.env import get_current_env
    site_dir = f"{paths['root']}/site_{site}"
    env = get_current_env()

    command = terraform_command_builder(command, site, args=args,
                                        options=options,
                                        needed_options=needed_options)

    env_vars = dict(os.environ)  # make a copy of the environment
    lp_key = 'LD_LIBRARY_PATH'  # for Linux and *BSD.
    lp_orig = env_vars.get(lp_key + '_ORIG')  # pyinstaller has this
    if lp_orig is not None:
        env_vars[lp_key] = lp_orig  # restore the original, unmodified value
    else:
        env_vars.pop(lp_key, None)  # last resort: remove the env var

    prev_sigint_handler = signal.getsignal(signal.SIGINT)
    try:
        p = subprocess.Popen(args=command, cwd=site_dir, env=env_vars)
        signal.signal(signal.SIGINT, signal.SIG_IGN)  # ignore on python thread
        p.wait()

        if p.returncode != 0:
            print(Fore.RED + "There was an error while running the terraform"
                  " command.")
            exit(1)
    except subprocess.CalledProcessError:
        print(Fore.RED + f"There was an error while doing the {command}"
              f" command inside the '{site}' site in '{env}' environment.")
        exit(1)
    finally:
        signal.signal(signal.SIGINT, prev_sigint_handler)


def terraform_destroy_all_sites():
    """
    Destroy all sites by looping through all sites in reverse order
    """
    sites = list(reversed(get_settings()["sites"]["ordered"]))
    for site in sites:
        terraform_command_runner("destroy", site, options=["-auto-approve"],
                                 needed_options="var")


def terraform_apply_all_sites():
    """
    Destroy all sites by looping through all sites
    """
    sites = list(get_settings()["sites"]["ordered"])
    for site in sites:
        if site != "inception":
            terraform_command_runner("init", site, options=["-reconfigure"],
                                     needed_options="backend")
        terraform_command_runner("apply", site, options=["-auto-approve"],
                                 needed_options="var")


def terraform_init_all_sites():
    """
    Init all sites by looping through all sites (inception doesn't need a init)
    """
    sites = list(get_settings()["sites"]["ordered"])
    if "inception" in sites:
        sites.remove("inception")
    for site in sites:
        print(Fore.YELLOW + f"Initializing {site} site")
        terraform_command_runner("init", site, options=["-reconfigure"],
                                 needed_options="backend")

File: terrabutler/tf.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"slices"
	"strings"
	"syscall"

	"go.uber.org/zap"
)

// Used for generate-options, prints arguments
func terraform_args_print(command string, site string) string {
	var needed_options string
	if command == "init" {
		needed_options = "backend"
	} else if command == "plan" || command == "apply" {
		needed_options = "var"
	} else {
		needed_options = ""
	}

	options := terraform_needed_options_builder(needed_options, site)
	return strings.Join(options, " ")
}

// Create array of needed options for backend or var files
func terraform_needed_options_builder(needed_options string, site string) []string {
	env := get_current_env()
	org := settings.String("general.organization")
	default_env := settings.String("environments.default.name")

	if needed_options == "backend" {
		backend_dir := paths["backends"]

		if site == "inception" { //Init inception with default ENV
			return []string{"-backend-config", backend_dir + "/" + org + "-" + default_env + "-inception.tfvars"}
		} else {
			return []string{"-backend-config", backend_dir + "/" + org + "-" + env + "-" + site + ".tfvars"}
		}
	} else if needed_options == "var" {
		variables_dir := paths["variables"]

		return []string{"-var-file", variables_dir + "/global.tfvars",
			"-var-file", variables_dir + "/" + org + "-" + env + ".tfvars",
			"-var-file", variables_dir + "/" + org + "-" + env + "-" + site + ".tfvars"}

	} else { // If needed_options is empty, return empty slice
		return []string{}
	}
}

// Command builder
func terraform_command_builder(command string, site string, args []string, options []string, needed_options string) []string {

	base_command := []string{"terraform"}
	base_command = append(base_command, strings.Split(command, " ")...)

	if needed_options == "backend" || needed_options == "var" {
		aux := terraform_needed_options_builder(needed_options, site)
		base_command = append(base_command, aux...)
	}

	base_command = append(base_command, options...)
	base_command = append(base_command, args...)

	return base_command
}

// Main runner function
func terraform_command_runner(command string, site string, args []string, options []string, needed_options string) {

	env := get_current_env()

	tf_binary, err := exec.LookPath("terraform")
	if err != nil {
		logger.Error("No Terraform executable found.")
		os.Exit(1)
	}

	//Changes the current working dir to the site chosen
	err = os.Chdir(paths["root"] + "/site_" + site)
	//In theory this error shouldn't occur, since is begin parsed before the execution of this command
	if err != nil {
		logger.Error("Error in finding the path for the site " + site)
		os.Exit(1)
	}

	runner_command := terraform_command_builder(command, site, args, options, needed_options)

	logger.Debug(fmt.Sprintf("Executing command: %v", runner_command))

	runner_env := os.Environ()

	logger.Debug("Tf Binary loc: " + tf_binary)

	execErr := syscall.Exec(tf_binary, runner_command, runner_env)
	if execErr != nil {
		logger.Error("There was an error during execution of terraform "+command+" in the site "+site+" in the environment "+env, zap.Error(execErr))
		os.Exit(1)
	}

}

// New commands to be used in all sites
func tf_destroy_all_sites() {
	sites := settings.Strings("sites.ordered")
	slices.Reverse(sites)
	for _, site := range sites {
		command := terraform_command_builder("destroy", site, []string{}, []string{"-auto-approve"}, "var")

		//Changes the current working dir to the site chosen
		err := os.Chdir(paths["root"] + "/site_" + site)
		if err != nil {
			logger.Error("Error in finding the path for the site " + site)
			os.Exit(1)
		}
		cmd := exec.Command(command[0], command[1:]...)
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		err = cmd.Run()
		if err != nil {
			logger.Error("There was an error during execution of terraform "+command[0]+"in the site "+site+" in the environment "+get_current_env(), zap.Error(err))
			os.Exit(1)
		}
	}

}

func tf_apply_all_sites() {
	sites := settings.Strings("sites.ordered")
	for _, site := range sites {
		if site != "inception" {
			command := terraform_command_builder("init", site, []string{}, []string{"-reconfigure"}, "backend")
			cmd := exec.Command(command[0], command[1:]...)
			cmd.Run()
		}
		command := terraform_command_builder("apply", site, []string{}, []string{"-auto-approve"}, "var")

		//Changes the current working dir to the site chosen
		err := os.Chdir(paths["root"] + "/site_" + site)
		if err != nil {
			logger.Error("Error in finding the path for the site " + site)
			os.Exit(1)
		}
		cmd := exec.Command(command[0], command[1:]...)
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		err = cmd.Run()
		if err != nil {
			logger.Error("There was an error during execution of terraform "+command[0]+"in the site "+site+" in the environment "+get_current_env(), zap.Error(err))
			os.Exit(1)
		}
	}
}

func tf_init_all_sites() {
	sites := settings.Strings("sites.ordered")
	if index := slices.Index(sites, "inception"); index != -1 {
		sites = sites[index+1:]
	}
	for _, site := range sites {

		logger.Warn("Initializing " + site + " site")
		command := terraform_command_builder("init", site, []string{}, []string{"-reconfigure"}, "backend")

		//Changes the current working dir to the site chosen
		err := os.Chdir(paths["root"] + "/site_" + site)
		if err != nil {
			logger.Error("Error in finding the path for the site " + site)
			os.Exit(1)
		}
		cmd := exec.Command(command[0], command[1:]...)
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr
		err = cmd.Run()
		if err != nil {
			logger.Error("There was an error during execution of terraform "+command[0]+"in the site "+site+" in the environment "+get_current_env(), zap.Error(err))
			os.Exit(1)
		}
	}
}
